### PR TITLE
[SEMI-MODULAR] EORG teleport

### PR DESCRIPTION
**Pull request by Kush1Push1**
## About The Pull Request

Added option in game preferences, which in toggled state teleports players after emergency shuttle's docking to Central Command in the thunderdome.

### How it works

1) Emergency shuttle undocks from the station. The game checks if EORG config is toggled on (which can be changed in config file or by admins in server tab). 
Players will recieve a message about coming round-end event and their current status of EORG preference. If a text at end of the message will be clicked, their status will be changed (be/not be at EORG).
2) After the shuttle's docking with CC, there will be the teleportation of all players. If the EORG config is still on and a player still has the preference, they will be moved to thunderdome.
3) All wounds and restrains will be removed.
4) Their gear will be deleted and replaced with one of random presets (so there is no point to prepare for EORG).
5) All EORG banned players are still in rotation to marked zone.

## How This Contributes To The Skyrat Roleplay Experience

EORG can be a thing for players who wish to have it. After all, everyone want to fight in this game from time to time. Anyone else will have their round end at central command/their respected areas.

## Changelog

:cl: SishTis
add: Added EORG teleportation option in character setup's game preferences. In toggled state it teleports a player to thunderdome after round end. Admins also have a button in Server's tab to disable it for some events.
/:cl:

### CODE REVIEW

File: code/modules/shuttle/emergency.dm [hunk 1]
```diff
@@ -509,7 +509,10 @@
 				setTimer(SSshuttle.emergency_escape_time * engine_coeff)
 				priority_announce("The Emergency Shuttle has left the station. Estimate [timeLeft(600)] minutes until the shuttle docks at Central Command.", null, ANNOUNCER_SHUTTLELEFT, "Priority") //SKYRAT EDIT CHANGE - ANNOUNCER_SHUTTLELEFT
 				INVOKE_ASYNC(SSticker, /datum/controller/subsystem/ticker.proc/poll_hearts)
-				bolt_all_doors() //SKYRAT EDIT ADDITION
+				//SKYRAT ADDITON BEGIN
```
Comment (twilightwanderer) [inline code/modules/shuttle/emergency.dm:512]: It is better not to forget to specify a link to the module.
```
bolt_all_doors() //SKYRAT EDIT ADDITION
//SKYRAT EDIT ADDITION BEGIN - EORG MODULE
INVOKE_ASYNC(GLOBAL_PROC, /proc/process_eorg_announce)
//SKYRAT EDIT ADDITION END - EORG MODULE
```

File: modular_skyrat/modules/eorg/eorg.dm [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		if(eorg_candidate.client)
+			var/eorg_choice = eorg_candidate.client.prefs?.read_preference(/datum/preference/toggle/eorg_join)
+			to_chat(eorg_candidate, "The round will end up soon. You have choose to [eorg_choice ? "join in" : "stay away"] before. \
+			<a href='?src=[REF(eorg_candidate)];eorg_teleport_switch=1'>Shall you change your mind?</a>")
+
+/proc/process_eorg_teleport()
+	if(!(CONFIG_GET(flag/allow_eorg)))
+		return
+	var/list/EORG_teleports = list()
+	var/list/EORG_outfits = list(\
```
Comment (Zonespace27) [inline modular_skyrat/modules/eorg/eorg.dm:36]: make these static lists

File: modular_skyrat/modules/eorg/eorg.dm [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+		if(!EORG_teleports) //empty list, for some reason
+			EORG_teleports += eorg_zone.loc
+
+		//temporary before a real map will be made
+		for(var/obj/machinery/door/poddoor/door in eorg_zone)
+			INVOKE_ASYNC(door, /obj/machinery/door/poddoor.proc/open)
+
+	for(var/mob/living/carbon/human/eorg_player in GLOB.human_list)
+		if(eorg_player.ckey)
```
Comment (Zonespace27) [inline modular_skyrat/modules/eorg/eorg.dm:58]: early continue

File: modular_skyrat/modules/eorg/eorg.dm [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+		var/eorg_choice = client.prefs.read_preference(/datum/preference/toggle/eorg_join)
+		to_chat(src, "You choose to [eorg_choice ?  "end the round in peace." : "fight. Prepare for blood!"]")
+		preferences.write_preference(GLOB.preference_entries[/datum/preference/toggle/eorg_join], !eorg_choice)
+		return
+
+/proc/process_eorg_announce()
+	if(!(CONFIG_GET(flag/allow_eorg)))
+		return
+	for(var/mob/living/carbon/human/eorg_candidate in GLOB.human_list)
+		if(eorg_candidate.client)
```
Comment (Zonespace27) [inline modular_skyrat/modules/eorg/eorg.dm:27]: early continue

File: modular_skyrat/modules/eorg/eorg.dm [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+/mob/living/carbon/human/Topic(href, href_list)
+	..()
+
+	if(href_list["eorg_teleport_switch"])
+		var/datum/preferences/preferences = client.prefs
+		var/eorg_choice = client.prefs.read_preference(/datum/preference/toggle/eorg_join)
+		to_chat(src, "You choose to [eorg_choice ?  "end the round in peace." : "fight. Prepare for blood!"]")
+		preferences.write_preference(GLOB.preference_entries[/datum/preference/toggle/eorg_join], !eorg_choice)
+		return
```
Comment (Zonespace27) [inline modular_skyrat/modules/eorg/eorg.dm:21]: does nothing

File: modular_skyrat/modules/eorg/eorg.dm [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+		if(!EORG_teleports) //empty list, for some reason
+			EORG_teleports += eorg_zone.loc
+
+		//temporary before a real map will be made
+		for(var/obj/machinery/door/poddoor/door in eorg_zone)
+			INVOKE_ASYNC(door, /obj/machinery/door/poddoor.proc/open)
+
+	for(var/mob/living/carbon/human/eorg_player in GLOB.human_list)
+		if(eorg_player.ckey)
+			if(eorg_player.client?.prefs?.read_preference(/datum/preference/toggle/eorg_join))
```
Comment (Zonespace27) [inline modular_skyrat/modules/eorg/eorg.dm:59]: early continue

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/skyrat/eorg_join.tsx [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { CheckboxInput, FeatureToggle } from '../../base';
+
+export const eorg_join: FeatureToggle = {
+  name: 'Round End Grief participating',
```
Comment (Zonespace27) [inline tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/skyrat/eorg_join.tsx:4]:
```suggestion
  name: 'End Of Round Grief Participation',
```

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

**Comment by Iamgoofball** [comment]
Why not just sign them up for CTF instead?

**Comment by weesnaw-ops** [comment]
> Why not just sign them up for CTF instead?

while i would normally agree ctf just isnt the same as beating the shit out of people in a bloody mess

**Comment by Aether217** [comment]
It has been discussed in the past and the thunderdome will remain for staff to test things only.

Also, this more or less does nothing than bring back staff having to police eorg prep again.

**Comment by Gandalf2k15** [comment]
No, we will not be having any EORG teleporting either. That's what CTF is for.

**Comment by ArdentSpark** [comment]
They don't have to police EORG prep because their gear is deleted
[Their gear will be deleted and replaced with one of random presets (so there is no point to prepare for EORG).]

**Comment by theselfish** [comment]
> Also, this more or less does nothing than bring back staff having to police eorg prep again.

image

**Comment by GuiltyNeko** [comment]
If they got teleported to CTF instead of the thunderdome, it'd be less bad and solve the "thunderdome is for admins" issue, and if their gear is deleted as mentioned, prepping means nothing.

**Comment by YakumoChen** [comment]
just make a CTF button you can press at round end that automatically signs you up and teleports you to one of the random CTF beacons when it begins

**Comment by Kush1Push1** [comment]
I'll fix the issues and reopen it.

**Comment by hugmeorelse** [comment]
@Zonespace27 This should have been closed and rejected as per discussion.

**Comment by hugmeorelse** [comment]
@Zonespace27 I believe you should have notified staff before requesting changes, because this PR would have never been accepted.

**Pull request closed without merging**